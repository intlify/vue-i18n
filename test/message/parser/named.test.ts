import {
  createParser,
  NodeTypes,
  MessageNode
} from '../../../src/message/parser'

let spy
beforeEach(() => {
  spy = jest.fn()
})

describe('basic', () => {
  test(`hi {name} !`, () => {
    const text = `hi {name} !`
    const parser = createParser({ onError: spy })
    const ast = parser.parse(text)

    expect(ast).toMatchSnapshot()
    expect(spy).not.toHaveBeenCalled()

    expect(ast.type).toEqual(NodeTypes.Resource)
    expect(ast.body.type).toEqual(NodeTypes.Message)
    const message = ast.body as MessageNode
    expect(message.items).toHaveLength(3)
    expect(message.items).toMatchObject([
      {
        type: NodeTypes.Text,
        value: 'hi '
      },
      {
        type: NodeTypes.Named,
        key: 'name'
      },
      {
        type: NodeTypes.Text,
        value: ' !'
      }
    ])
  })
})

describe('multiple', () => {
  /* eslint-disable no-irregular-whitespace */
  test(`{first} {middle}　{last}`, () => {
    const text = `{first} {middle}　{last}`
    const parser = createParser({ onError: spy })
    const ast = parser.parse(text)

    expect(ast).toMatchSnapshot()
    expect(spy).not.toHaveBeenCalled()

    expect(ast.type).toEqual(NodeTypes.Resource)
    expect(ast.body.type).toEqual(NodeTypes.Message)
    const message = ast.body as MessageNode
    expect(message.items).toHaveLength(5)
    expect(message.items).toMatchObject([
      {
        type: NodeTypes.Named,
        key: 'first'
      },
      {
        type: NodeTypes.Text,
        value: ' '
      },
      {
        type: NodeTypes.Named,
        key: 'middle'
      },
      {
        type: NodeTypes.Text,
        value: '　'
      },
      {
        type: NodeTypes.Named,
        key: 'last'
      }
    ])
  })
  /* eslint-enable no-irregular-whitespace */
})

describe('whitespaces in placeholder', () => {
  test(`hi {  name } !`, () => {
    const text = `hi {  name } !`
    const parser = createParser({ onError: spy })
    const ast = parser.parse(text)

    expect(ast).toMatchSnapshot()
    expect(spy).not.toHaveBeenCalled()

    expect(ast.type).toEqual(NodeTypes.Resource)
    expect(ast.body.type).toEqual(NodeTypes.Message)
    const message = ast.body as MessageNode
    expect(message.items).toHaveLength(3)
    expect(message.items).toMatchObject([
      {
        type: NodeTypes.Text,
        value: 'hi '
      },
      {
        type: NodeTypes.Named,
        key: 'name'
      },
      {
        type: NodeTypes.Text,
        value: ' !'
      }
    ])
  })
})

describe('new lines in placeholder', () => {
  test(`{first}\n{middle}\n{last}`, () => {
    const text = `{first}\n{middle}\n{last}`
    const parser = createParser({ onError: spy })
    const ast = parser.parse(text)

    expect(ast).toMatchSnapshot()
    expect(spy).not.toHaveBeenCalled()

    expect(ast.type).toEqual(NodeTypes.Resource)
    expect(ast.body.type).toEqual(NodeTypes.Message)
    const message = ast.body as MessageNode
    expect(message.items).toHaveLength(5)
    expect(message.items).toMatchObject([
      {
        type: NodeTypes.Named,
        key: 'first'
      },
      {
        type: NodeTypes.Text,
        value: '\n'
      },
      {
        type: NodeTypes.Named,
        key: 'middle'
      },
      {
        type: NodeTypes.Text,
        value: '\n'
      },
      {
        type: NodeTypes.Named,
        key: 'last'
      }
    ])
  })
})

test.todo(`hi {} !`)
test.todo(`hi {{}} !`)
test.todo(`hi {  } !`)
test.todo(`hi {name$} !`)
test.todo(`hi {snake_case} !`)
test.todo(`hi {\nname\n} !`)
test.todo(`hi {{name}} !`)
test.todo(`hi { { name } } !`)
test.todo(`hi {name !`)
test.todo(`hi { name !`)
test.todo(`hi {  name !`)
test.todo(`hi {@.lower:name !`)
test.todo(`hi { @:name !`)
test.todo(`hi {  | hello {name} !`)
test.todo(`hi { @:name  | hello {name} !`)
