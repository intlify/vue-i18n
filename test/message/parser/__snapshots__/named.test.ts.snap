// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic hi {name} ! 1`] = `
Object {
  "body": Object {
    "end": 11,
    "items": Array [
      Object {
        "end": 3,
        "loc": Object {
          "end": Object {
            "column": 4,
            "line": 1,
            "offset": 3,
          },
          "start": Object {
            "column": 1,
            "line": 1,
            "offset": 0,
          },
        },
        "start": 0,
        "type": 3,
        "value": "hi ",
      },
      Object {
        "end": 9,
        "key": "name",
        "loc": Object {
          "end": Object {
            "column": 10,
            "line": 1,
            "offset": 9,
          },
          "start": Object {
            "column": 4,
            "line": 1,
            "offset": 3,
          },
        },
        "start": 3,
        "type": 4,
      },
      Object {
        "end": 11,
        "loc": Object {
          "end": Object {
            "column": 12,
            "line": 1,
            "offset": 11,
          },
          "start": Object {
            "column": 10,
            "line": 1,
            "offset": 9,
          },
        },
        "start": 9,
        "type": 3,
        "value": " !",
      },
    ],
    "loc": Object {
      "end": Object {
        "column": 12,
        "line": 1,
        "offset": 11,
      },
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "start": 0,
    "type": 2,
  },
  "end": 11,
  "loc": Object {
    "end": Object {
      "column": 12,
      "line": 1,
      "offset": 11,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "start": 0,
  "type": 0,
}
`;

exports[`multiple {first} {middle}　{last} 1`] = `
Object {
  "body": Object {
    "end": 23,
    "items": Array [
      Object {
        "end": 7,
        "key": "first",
        "loc": Object {
          "end": Object {
            "column": 8,
            "line": 1,
            "offset": 7,
          },
          "start": Object {
            "column": 1,
            "line": 1,
            "offset": 0,
          },
        },
        "start": 0,
        "type": 4,
      },
      Object {
        "end": 8,
        "loc": Object {
          "end": Object {
            "column": 9,
            "line": 1,
            "offset": 8,
          },
          "start": Object {
            "column": 8,
            "line": 1,
            "offset": 7,
          },
        },
        "start": 7,
        "type": 3,
        "value": " ",
      },
      Object {
        "end": 16,
        "key": "middle",
        "loc": Object {
          "end": Object {
            "column": 17,
            "line": 1,
            "offset": 16,
          },
          "start": Object {
            "column": 9,
            "line": 1,
            "offset": 8,
          },
        },
        "start": 8,
        "type": 4,
      },
      Object {
        "end": 17,
        "loc": Object {
          "end": Object {
            "column": 18,
            "line": 1,
            "offset": 17,
          },
          "start": Object {
            "column": 17,
            "line": 1,
            "offset": 16,
          },
        },
        "start": 16,
        "type": 3,
        "value": "　",
      },
      Object {
        "end": 23,
        "key": "last",
        "loc": Object {
          "end": Object {
            "column": 24,
            "line": 1,
            "offset": 23,
          },
          "start": Object {
            "column": 18,
            "line": 1,
            "offset": 17,
          },
        },
        "start": 17,
        "type": 4,
      },
    ],
    "loc": Object {
      "end": Object {
        "column": 24,
        "line": 1,
        "offset": 23,
      },
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "start": 0,
    "type": 2,
  },
  "end": 23,
  "loc": Object {
    "end": Object {
      "column": 24,
      "line": 1,
      "offset": 23,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "start": 0,
  "type": 0,
}
`;

exports[`new lines in placeholder {first}
{middle}
{last} 1`] = `
Object {
  "body": Object {
    "end": 24,
    "items": Array [
      Object {
        "end": 7,
        "key": "first",
        "loc": Object {
          "end": Object {
            "column": 8,
            "line": 1,
            "offset": 7,
          },
          "start": Object {
            "column": 1,
            "line": 1,
            "offset": 0,
          },
        },
        "start": 0,
        "type": 4,
      },
      Object {
        "end": 8,
        "loc": Object {
          "end": Object {
            "column": 1,
            "line": 2,
            "offset": 8,
          },
          "start": Object {
            "column": 8,
            "line": 1,
            "offset": 7,
          },
        },
        "start": 7,
        "type": 3,
        "value": "
",
      },
      Object {
        "end": 16,
        "key": "middle",
        "loc": Object {
          "end": Object {
            "column": 9,
            "line": 2,
            "offset": 16,
          },
          "start": Object {
            "column": 1,
            "line": 2,
            "offset": 8,
          },
        },
        "start": 8,
        "type": 4,
      },
      Object {
        "end": 18,
        "loc": Object {
          "end": Object {
            "column": 1,
            "line": 3,
            "offset": 18,
          },
          "start": Object {
            "column": 9,
            "line": 2,
            "offset": 16,
          },
        },
        "start": 16,
        "type": 3,
        "value": "
",
      },
      Object {
        "end": 24,
        "key": "last",
        "loc": Object {
          "end": Object {
            "column": 7,
            "line": 3,
            "offset": 24,
          },
          "start": Object {
            "column": 1,
            "line": 3,
            "offset": 18,
          },
        },
        "start": 18,
        "type": 4,
      },
    ],
    "loc": Object {
      "end": Object {
        "column": 7,
        "line": 3,
        "offset": 24,
      },
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "start": 0,
    "type": 2,
  },
  "end": 24,
  "loc": Object {
    "end": Object {
      "column": 7,
      "line": 3,
      "offset": 24,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "start": 0,
  "type": 0,
}
`;

exports[`whitespaces in placeholder hi {  name } ! 1`] = `
Object {
  "body": Object {
    "end": 14,
    "items": Array [
      Object {
        "end": 3,
        "loc": Object {
          "end": Object {
            "column": 4,
            "line": 1,
            "offset": 3,
          },
          "start": Object {
            "column": 1,
            "line": 1,
            "offset": 0,
          },
        },
        "start": 0,
        "type": 3,
        "value": "hi ",
      },
      Object {
        "end": 12,
        "key": "name",
        "loc": Object {
          "end": Object {
            "column": 13,
            "line": 1,
            "offset": 12,
          },
          "start": Object {
            "column": 4,
            "line": 1,
            "offset": 3,
          },
        },
        "start": 3,
        "type": 4,
      },
      Object {
        "end": 14,
        "loc": Object {
          "end": Object {
            "column": 15,
            "line": 1,
            "offset": 14,
          },
          "start": Object {
            "column": 13,
            "line": 1,
            "offset": 12,
          },
        },
        "start": 12,
        "type": 3,
        "value": " !",
      },
    ],
    "loc": Object {
      "end": Object {
        "column": 15,
        "line": 1,
        "offset": 14,
      },
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "start": 0,
    "type": 2,
  },
  "end": 14,
  "loc": Object {
    "end": Object {
      "column": 15,
      "line": 1,
      "offset": 14,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "start": 0,
  "type": 0,
}
`;
